src/lattices/abstract.jl

```diff
@@ -0,0 +1,62 @@
+"""
+    AbstractLattice
+
+AbstractLattice unifies lattices defined in the MonteCarlo package. As such it
+defines standard methods for `length(l)` and `neighbors(l)` where
+`l <: AbstractLattice`.
+"""
+abstract type AbstractLattice end
+
+
+# Number of sites
+Base.length(l::AbstractLattice) = l.sites
+
+"""
+    neighbors(l::AbstractLattice, drop_repeats::Val{true})
+    neighbors(l::AbstractLattice, drop_repeats::Val{false})
+
+Returns an iterator over bonds, given as tuples (source index, target index). If
+`drop_repeats = Val(falsee)` bonds are assumed to be directed, i.e. both
+`(1, 2)` and `(2, 1)` are included. If `drop_repeats = Val(true)` bonds are
+assumed to be undirected, i.e. `(1, 2)` and `(2, 1)` are assumed to be
+equivalent and only one of them will be counted.
+"""
+function neighbors(l::AbstractLattice, drop_repeats::Val{false})
+    (
+        [source, target] for source in 1:length(l)
+            for target in l.neighs[:, source]
+    )
+end
+function neighbors(l::AbstractLattice, drop_repeats::Val{true})
+    (
+        l.bonds[i, 1:2] for i in 1:size(l.bonds, 1)
+    )
+end
+
+
+
+
+"""
+    AbstractCubicLattice
+
+AbstractCubicLattice includes 1D Chains, 2D square lattices and ND cubic
+lattices.
+"""
+abstract type AbstractCubicLattice <: AbstractLattice end
+
+
+
+
+# LatPhys
+
+Base.length(l::LatPhysBase.AbstractLattice) = LatPhysBase.numSites(l)
+function neighbors(l::LatPhysBase.AbstractLattice, drop_repeats::Val{false})
+    ((LatPhysBase.from(b), LatPhysBase.to(b)) for b in LatPhysBase.bonds(l))
+end
+function neighbors(l::LatPhysBase.AbstractLattice, drop_repeats::Val{true})
+    (
+        (LatPhysBase.from(b), LatPhysBase.to(b))
+        for b in LatPhysBase.bonds(l)
+        if LatPhysBase.from(b) < LatPhysBase.to(b)
+    )
+end
```

Comment:
As a general note, keep in mind that we certainly want to be able to initially create a static lookup table of neighbors in MC to get O(1) lookup.

Reply:
I don't think LatPhys has something like that. So we should add a thin wrapper which generates the `neighs` from MonteCarlo.jl, containing the same information as `neighbors(lattice, Val(false))`. I don't think we need one for `neighbors(lattice, Val(true))` though.

Imo it would be useful to have
* `all_neighbors(lattice, site)` -> Vector of targets
* `no_repeated_neighbors(lattice)` -> Iterable of (source, target) pairs

I don't think there is a reason to keep an `all_neighbors(lattice)` around.

Reply:
I added a wrapper for LatPhys lattices and a method `neighbors(lattice, site)`

Reply:
One more thing: do we explicitly want to depend on LatPhys*.jl (in Project.toml)? Maybe it's better to use Requires.jl to only load the LatticePhysics stuff when, say, LatPhysBase.jl has been explicitly loaded.

Reply:
I think that's a good idea. I'll make another commit for this.

Btw, do you want me to remove the Honeycomb lattice from this?

Reply:
I guess it doesn't hurt to leave it in there for now. We can remove it later if you want/have to. BTW: I'm currently revisiting MonteCarlo.jl on a general (somewhat higher) level. So don't wonder about some changes leading to merge conflicts. We'll fix them later.